lib/Doctrine/DBAL/Driver/OCI8/OCI8Exception.php

```diff
@@ -17,6 +19,6 @@ public static function fromErrorInfo($error)
             return new self('Database error occurred but no error information was retrieved from the driver.');
         }
 
-        return new self($error['message'], null, $error['code']);
+        return new self($error['message'] ?? 'An unknown error occurred.', null, $error['code'] ?? 0);
```

Comment:
Is it really possible that the error is not `false` but doesn't have a message or code?

Reply:
IIRC the issue was with message being null.

Reply:
It's `null` because it's mocked, e.g. in `OCI8StatementTest::testExecute()`. `OCI8Statement::errorInfo()` replaces `false` from the driver with `[]` for DBAL API compliance and therefore cannot be used as input for `OCI8Exception::fromErrorInfo()` which expects `false` as the indicator of no error.

Please try replacing all occurrences of `throw OCI8Exception::fromErrorInfo($this->errorInfo())` in `OCI8Statement` with `throw OCI8Exception::fromErrorInfo(oci_error($this->_sth))`.

I'd like to see `Statement::errorInfo()` and `::errorCode()` gone from the API entirely in the future.

Reply:
The problem is that oci_error() returns array|false...

Reply:
Another issue is that some calls to `oci_error()` are in context when no _dbh_ is available, so suggested replacement won't work. :|

Reply:
> The problem is that oci_error() returns array|false...

It's not a problem, this is exactly what I'm looking for.

> Another issue is that some calls to `oci_error()` are in context when no _dbh_ is available, so suggested replacement won't work. :|

It workes just fine. As long `$_sth` is `protected`, it's accessible from the mock, and `oci_error()` doesn't complain about getting `null` instead of a resource. It happily returns `false`.
```diff
--- a/lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php
+++ b/lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php
@@ -298,7 +298,7 @@ class OCI8Statement implements IteratorAggregate, Statement
             $length ?? -1,
             $this->convertParameterType($type)
         )) {
-            throw OCI8Exception::fromErrorInfo($this->errorInfo());
+            throw OCI8Exception::fromErrorInfo(oci_error($this->_sth));
         }
     }
 
@@ -390,7 +390,7 @@ class OCI8Statement implements IteratorAggregate, Statement
 
         $ret = @oci_execute($this->_sth, $this->_conn->getExecuteMode());
         if (! $ret) {
-            throw OCI8Exception::fromErrorInfo($this->errorInfo());
+            throw OCI8Exception::fromErrorInfo(oci_error($this->_sth));
         }
 
         $this->result = true;
```